espnet2/enh/separator/ineube_separator.py

```diff
@@ -0,0 +1,147 @@
+        if window == "hanning":
+            window = torch.hann_window(n_fft)
+        elif window == "hamming":
+            window = torch.hamming_window(n_fft)
+        elif window is None:
+            pass
+        else:
+            raise NotImplementedError
+        fft_c_channels = n_fft // 2 + 1
+        self.enc, self.dec = make_enc_dec(
+            "torch_stft", n_fft, n_fft, stride, window=window
+        )
+        self.dnn1 = TCNDenseUNet(
+            n_spk,
+            fft_c_channels,
+            mic_channels,
+            hid_chans,
+            hid_chans_dense,
+            ksz_dense,
+            ksz_tcn,
+            tcn_repeats,
+            tcn_blocks,
+            tcn_channels,
+            activation=activation,
+        )
+
+        self.dnn2 = TCNDenseUNet(
+            n_spk,
+            fft_c_channels,
+            mic_channels + 2,
+            hid_chans,
+            hid_chans_dense,
+            ksz_dense,
+            ksz_tcn,
+            tcn_repeats,
+            tcn_blocks,
+            tcn_channels,
+            activation=activation,
+        )
+
+    def unfold(self, tf_rep, chunk_size):
+        bsz, freq, _ = tf_rep.shape
+
+        est_unfolded = torch.nn.functional.unfold(
+            torch.nn.functional.pad(
+                tf_rep, (chunk_size, chunk_size), mode="constant"
+            ).unsqueeze(-1),
+            kernel_size=(2 * chunk_size + 1, 1),
+            padding=(0, 0),
+            stride=(1, 1),
+        )
+        n_chunks = est_unfolded.shape[-1]
+        est_unfolded = est_unfolded.reshape(bsz, freq, 2 * chunk_size + 1, n_chunks)
+        est_unfolded = est_unfolded.transpose(1, 2)
+        return est_unfolded
+
+    def mfmcwf(self, mixture, estimate, n_chunks):
+        bsz, mics, _, frames = mixture.shape
+        mix_unfolded = self.unfold(
+            mixture.reshape(bsz * mics, -1, frames), n_chunks
+        ).reshape(bsz, mics * (2 * n_chunks + 1), -1, frames)
+
+        mix_unfolded = af_transforms.to_torch_complex(mix_unfolded.double())
+        estimate1 = af_transforms.to_torch_complex(estimate.double())
+        zeta = torch.einsum("bmft, bft->bmf", mix_unfolded, estimate1.conj())
+        scm_mix = torch.einsum("bmft, bnft->bmnf", mix_unfolded, mix_unfolded.conj())
+        inv_scm_mix = torch.inverse(
+            condition_scm(scm_mix.permute(0, 3, 1, 2), 1e-4)
+        ).permute(0, 2, 3, 1)
+        bf_vector = torch.einsum("bmnf, bnf->bmf", inv_scm_mix, zeta)
+
+        beamformed = torch.einsum("...mf,...mft->...ft", bf_vector.conj(), mix_unfolded)
+        beamformed = af_transforms.from_torch_complex(beamformed).float()
+
+        return beamformed
+
+    def forward(
+        self,
+        input: Union[torch.Tensor, ComplexTensor],
+        ilens: torch.Tensor,
+        output_from="dnn1",
+        n_chunks=3,
+        freeze_dnn1=False,
+    ) -> Tuple[List[Union[torch.Tensor, ComplexTensor]], torch.Tensor, OrderedDict]:
+
+        mix_stft = self.enc(input)
```

Comment:
In my local trial, I encountered the following error:
```python
Traceback (most recent call last):
  File "espnet-0.10.7a1/tools/venv/envs/espnet/lib/python3.8/runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "espnet-0.10.7a1/tools/venv/envs/espnet/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "espnet-0.10.7a1/espnet2/bin/enh_train.py", line 23, in <module>
    main()
  File "espnet-0.10.7a1/espnet2/bin/enh_train.py", line 19, in main
    EnhancementTask.main(cmd=cmd)
  File "espnet-0.10.7a1/espnet2/tasks/abs_task.py", line 1019, in main
    cls.main_worker(args)
  File "espnet-0.10.7a1/espnet2/tasks/abs_task.py", line 1315, in main_worker
    cls.trainer.run(
  File "espnet-0.10.7a1/espnet2/train/trainer.py", line 286, in run
    all_steps_are_invalid = cls.train_one_epoch(
  File "espnet-0.10.7a1/espnet2/train/trainer.py", line 522, in train_one_epoch
    retval = model(**batch)
  File "espnet-0.10.7a1/tools/venv/envs/espnet/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1102, in _call_impl
    return forward_call(*input, **kwargs)
  File "espnet-0.10.7a1/espnet2/enh/espnet_model.py", line 160, in forward
    l, s, o = loss_wrapper(speech_ref, speech_pre, o)
  File "espnet-0.10.7a1/tools/venv/envs/espnet/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1102, in _call_impl
    return forward_call(*input, **kwargs)
  File "espnet-0.10.7a1/espnet2/enh/loss/wrappers/fixed_order.py", line 31, in forward
    loss += torch.mean(self.criterion(r, i)) / num_spk
  File "espnet-0.10.7a1/tools/venv/envs/espnet/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1102, in _call_impl
    return forward_call(*input, **kwargs)
  File "espnet-0.10.7a1/espnet2/enh/loss/criterions/time_domain.py", line 165, in forward
    target_stft = stft_enc(target)
  File "espnet-0.10.7a1/tools/venv/envs/espnet/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1102, in _call_impl
    return forward_call(*input, **kwargs)
  File "espnet-0.10.7a1/tools/venv/envs/espnet/lib/python3.8/site-packages/asteroid_filterbanks/enc_dec.py", line 177, in forward
    spec = multishape_conv1d(
  File "espnet-0.10.7a1/tools/venv/envs/espnet/lib/python3.8/site-packages/asteroid_filterbanks/scripting.py", line 37, in wrapper
    return fn(*args, **kwargs)
  File "espnet-0.10.7a1/tools/venv/envs/espnet/lib/python3.8/site-packages/asteroid_filterbanks/enc_dec.py", line 210, in multishape_conv1d
    return F.conv1d(waveform.unsqueeze(1), filters, stride=stride, padding=padding)
RuntimeError: Input type (torch.cuda.FloatTensor) and weight type (torch.FloatTensor) should be the same
```
Is it as expected?

Reply:
I am moving all the code to native ESPNet-SE now